### PR TITLE
Add code from rust backend

### PR DESCRIPTION
**Pull request by arjentz**

### CODE REVIEW

**Comment by rowanG077** [comment]
please ping me @arjenzijlstra once it's out of draft phase I will then look over it and merge.

**Comment by arjenzijlstra** [comment]
I wasn't able to get the front-end running, are there any updated instructions?

File: cryptify-back-end/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,379 @@
...
+        return Ok(None);
+    }
+
+    let start = headers.content_range.start.ok_or_else(|| Error::BadRequest(Some(
+        "Could not read Content-Range start".to_owned(),
+    )))?;
+    let end = headers.content_range.end.ok_or_else(|| Error::BadRequest(Some(
+        "Could not read Content-Range start".to_owned(),
+    )))?;
+    if end - start > 1024 * 1024 {
```
Comment (Iso5786) [inline cryptify-back-end/src/main.rs:256]: Maybe define the 1024 x 1024 as a const?

File: cryptify-back-end/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,379 @@
...
+    let mut file = match OpenOptions::new()
+        .write(true)
+        .open(Path::new(config.data_dir()).join(uuid))
+        .await
+    {
+        Ok(v) => v,
+        Err(_) => return Ok(None),
+    };
+
+    if start >= end || state.uploaded != start {
```
Comment (Iso5786) [inline cryptify-back-end/src/main.rs:277]: Move row 277 - 281 above file handling

File: cryptify-back-end/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,379 @@
...
+        return Err(Error::BadRequest(Some(
+            "Incorrect Content-Range header".to_owned(),
+        )));
+    }
+
+    file.seek(std::io::SeekFrom::Start(start))
+        .await
+        .map_err(|_| Error::InternalServerError(Some("Could not write file".to_owned())))?;
```
Comment (Iso5786) [inline cryptify-back-end/src/main.rs:285]: Couldn't you just use an .or instead of .map_err (as closure is unused)?

**Review by Iso5786 — CHANGES_REQUESTED** [review]

File: cryptify-back-end/src/main.rs (hunk 3 above)
Comment (arjentz) [reply cryptify-back-end/src/main.rs:285]: Yes, but it is optimised by Clippy by:
https://rust-lang.github.io/rust-clippy/master/index.html#or_fun_call
And then:
https://rust-lang.github.io/rust-clippy/master/index.html#bind_instead_of_map

**Comment by leonbotros** [comment]
Thanks @arjentz!

**Pull request merged**